NE-761: Support for admin configured CA trust bundle in Ingress Operator (#1514)
openshift/enhancements: enhancements/ingress/admin-configured-ca-bundle.md

see-also:
  - None
replaces:
  - None
superseded-by:
  - None
---

# Support for admin configured CA trust bundle in Ingress Operator

## Summary

This enhancement describes the proposal to allow an administrator of OpenShift to configure
trusted Certificate Authorities (CA) to verify the services' certificate by OpenShift Router
for the re-encrypted traffic.

## Motivation

When an OpenShift administrator creates a route with the re-encrypt termination type, they
have to also provide a CA certificate for that route if it uses a custom CA instead of the
default OpenShift CA for certificate validation. In addition, when the same CA is used by the
administrator for signing several services' certificates it is time-consuming and error-prone
to configure the same CA certificate in every created route object.
Likewise, we'd like to improve the process for updating a CA certificate when expired,
compromised or for any other reason in each route object. Instead, this document proposes a
simpler alternative: a feature in which an administrator could define a CA bundle for verifying
certificates from a ConfigMap consisting of the list of CA certificates, associated OpenShift

> Ingress Controller, and trusted OpenShift CA.

Reviewer: Note that there can be several different Ingress Controllers when the routes are sharded into subsets of routes.  i don't think this proposal covers multiple Ingress Controllers, so it should be pointed out as a non-goal.

> Ingress Controller, and trusted OpenShift CA.

Author: Sentences have got overlapped when I was apply the changes, updated it.

### User Stories
- As an OpenShift cluster administrator, I want to configure admin CA certificate bundle which
  should be used for verifying service's certificate by the OpenShift router, instead of
  having to configure the same as destinationCACertificate in each of the created route object.
- As an OpenShift developer, I want to create a CA trust bundle for Ingress operator usage,
  which includes the OpenShift service CA and the admin provided CA bundle as the default CA
  for verifying the service's certificate when a destination CA certificate is not configured

>   for a route.

Author: I meant to say, as a developer I want to add the functionality of creating a common CA bundle and using it as defaultDestinationCA to the operator which manages the OpenShift router. Please let me know what do you think.

>   for a route.

Reviewer 2: Oh, I see what you mean now!  Yeah, I've seen some other EPs with user stories that are written from the EP author's perspective.  In contrast, I always thought user stories were supposed to be written from the end-user's perspective.  Moreover, the entire EP itself presents the EP author's perspective, so a user story for the EP author seems superfluous.  However, if you think it is fine, I'll leave it be.

- As an OpenShift developer, I want to update the OpenShift router certificate cache whenever
  the OpenShift service CA certificate or the admin configured CA certificate bundle is updated

>   or deleted.

Reviewer: Add one more user story so that QE is aware this is still possible, and it is still tested:
As a developer, I want to be able to continue to override the default CA's by router annotation.

Reviewer: I would add an explicit story to ensure the ability to override an admin default is tested by QE.

### Goals

- An OpenShift administrator can configure a CA certificate bundle and expect that to be used
  for verifying the service's certificate by the OpenShift router without defining the
  destinationCACertificate in the route object created for the service.
- Openshift Router should be updated seamlessly when the OpenShift service CA certificate or
  the administrator configured CA certificate bundles are modified.

### Non-Goals

- OpenShift administrator defined CA certificate bundle will be used only for the traffic
  through OpenShift router and for the routes created with re-encrypt termination type, any
  other scenario is outside the scope of this proposal.

> 

Reviewer: This proposal covers the default Ingress Controller only, and does not apply to sharded Ingress Controllers.

Author: Yes, the same CA bundle is made available for all IngressControllers, but the implementation is dependent on the default IngressController for reconciling and cleanup of `ingress-ca-bundle` ConfigMap.

Reviewer 2: Why does the implementation need to watch any IngressController?  It seems to me that you need a new controller that only watches the service CA ConfigMap, custom CA ConfigMap, and `ingress-ca-bundle` ConfigMap.

Author: I had this misunderstanding other ingresscontroller existence won't matter when default ingresscontroller doesn't exist. 
Correction, implementation does not depend on a IngressController.

## Proposal

> OpenShift Ingress Operator to create a CA trust bundle which will contain the default OpenShift
> service CA certificate and the custom user-supplied CA certificates bundle and makes it available
> to the OpenShift Router as the default destination CA bundle file to be used for verifying the
> services' certificate. And also updates the CA bundle file whenever either of the certificates
> change, which should be watched by the OpenShift Router and update the same internally.
> 

Reviewer: ```suggestion
Change OpenShift Ingress Operator to:

1. Create a CA trust bundle which will contain the default OpenShift service CA certificate and a single custom user-supplied CA certificates bundle.
2. Make the new common certificate available to the OpenShift Router as the default destination CA bundle file to be used for verifying the services' certificate.
3. Updates the CA bundle file whenever either of the certificates change.
 
Update router to:

1. Watch the common CA bundle and update its configuration internally when needed.

```

### Workflow Description

> - An OpenShift administrator configures the list of trusted CA certificates as a ConfigMap.

Reviewer: ```suggestion
- An OpenShift administrator configures the list of trusted CA certificates as a ConfigMap, named specifically ________.
```

Author: suggestion incorporated.

> - OpenShift Ingress Operator creates a ConfigMap with the admin configured CA certificates and the

Reviewer: ```suggestion
- OpenShift Ingress Operator creates a ConfigMap by concatenating the admin configured CA certificates and the
```

  OpenShift service CA certificate stored in the respective ConfigMaps.
- OpenShift Ingress Operator creates a watcher and reconciles whenever changes are observed
  in the aforementioned ConfigMap objects.
- Openshift Router uses the operator created ConfigMap as the default destination CA certificate
  to verify services' certificate, when a route with re-encrypt termination type is created

>   without destinationCA configured.

Reviewer: ```suggestion
  with an empty destination CA.
```

- OpenShift router watches and reloads whenever CA certificate bundle changes.

![alt Workflow diagram](admin-configured-ca-bundle.png)

#### Variation [optional]
None

### API Extensions

> None

Author: comment incorporated.

### Implementation Details/Notes/Constraints [optional]

> An OpenShift administrator will have to create a ConfigMap with the fixed name `admin-ca-bundle`

Author: I thought to use the generic name with the intention, ca-bundle usage could be extended for any other similar use-cases as well.

> An OpenShift administrator will have to create a ConfigMap with the fixed name `admin-ca-bundle`

Reviewer 2: Hm, so it could be more general in the future, like a custom service CA?  @deads2k, I wonder whether you have an opinion on this.

> An OpenShift administrator will have to create a ConfigMap with the fixed name `admin-ca-bundle`

Reviewer 2: On further thought, would it make sense instead to define a label, and change cluster-ingress-operator to incorporate all CA certificates from all ConfigMaps with that label into the router's trust bundle?  

The reason I ask is because I was thinking about how something like cert-manager could use this feature, and it occurred to me that it would be much easier for independent simultaneous actors (humans or controllers) to use this feature to inject their respective certificates into the router's trust bundle if you used a well known label (allowing multiple ConfigMaps) instead of a well known name (requiring everything to be put into a single ConfigMap).

```yaml
apiVersion: v1
data:
  ca-bundle.crt: |
    -----BEGIN CERTIFICATE-----
    <service-ca>
    -----END CERTIFICATE-----
    -----BEGIN CERTIFICATE-----
    <admin-ca>
    -----END CERTIFICATE-----    
kind: ConfigMap
metadata:
  name: ingress-ca-bundle
  namespace: openshift-ingress
```

OpenShift Ingress operator should create a watcher to observe `admin-ca-bundle` and
`service-ca-bundle` for any changes and update `ingress-ca-bundle` accordingly.

OpenShift Router currently uses the CA certificate provided as a process argument
`default-destination-ca-path` or set as an environment variable `DEFAULT_DESTINATION_CA_PATH`
as the CA certificate for verifying the service's certificate for re-encrypt termination type.

CA certificate bundle is currently made available to the router using the volume mount option 
at the path `/var/run/configmaps/service-ca/service-ca.crt` which will be updated to 
`/var/run/configmaps/ca-trust/ca-bundle.crt` with the `ingress-ca-bundle` content.

OpenShift Router should create a file watcher to observe `/var/run/configmaps/ca-trust/ca-bundle.crt`
file and reload when the file content has been modified.

### Risks and Mitigations

> N/A.

Author: Updated `Risks and Mitigations` section.

### Drawbacks

> When `ingress-ca-bundle` ConfigMap contents made available to OpenShift Router as a file
> though volume mount is updated, certain [delay](https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#mounted-configmaps-are-updated-automatically) 

Reviewer: ```suggestion
When a user updates the `ingress-ca-bundle` ConfigMap, its contents are made available to OpenShift Router are eventually updated, though a certain [delay](https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#mounted-configmaps-are-updated-automatically) 
```

is expected for the new changes to be propagated to the pod.

## Design Details

### Open Questions [optional]

> N/A.

Reviewer: Please add that the solution is only for a non-sharded environment, in which there is only a default Ingress Controller.

Author: Non-Goals section updated to indicate the `ingress-ca-bundle` is common for all IngressControllers.

### Test Plan

- Create routes with re-encrypt termination type with and without destinationCA.

> - Upgrade/downgrade testing.

Author: comment incorporated.

> - Sufficient time for feedback.

Author: Updated to mean, sufficient time for feedback from QE team.

- End user documentation
- Sufficient test coverage
- Gather feedback from users rather than just developers

#### Tech Preview -> GA

- More testing (upgrade, downgrade, scale)
- Sufficient time for feedback
- Available by default
- Conduct load testing
- User facing documentation created in [openshift-docs](https://github.com/openshift/openshift-docs/)

#### Removing a deprecated feature
N/A.

### Upgrade / Downgrade Strategy

On upgrade:
- OpenShift Ingress Operator would have the functionality to look for `admin-ca-bundle` in
  `openshift-config` namespace and would create `ingress-ca-bundle` in `openshift-ingress`
  namespace and OpenShift Router would refer to the new CA bundle content.

On downgrade:
- OpenShift Ingress Operator created `ingress-ca-bundle` in `openshift-ingress` namespace
  might still exist, but it will not be used by the operator or the router. And the routes
  created post upgrade will become inaccessible, as the services' certificate validation would
  fail without a destinationCACertificate, until destinationCACertificate is configured for
  each inaccessible route.

### Version Skew Strategy

N/A.

### Operational Aspects of API Extensions

> N/A.

Author: comment incorporated.

#### Failure Modes

N/A.

#### Support Procedures

> N/A.

Author: section updated.

## Implementation History

N/A.

## Alternatives

> N/A.

Author: section updated.

---
title: admin-configured-ca-bundle
authors:
  - "@bhb"
reviewers:
  - "@Miciah" ## reviewer for cluster-ingress-operator component
  - "@tgeer" ## reviewer for CFE
approvers:
  - "@jerpeter1" ## approver for CFE
  - "@Miciah" ## approver for cluster-ingress-operator component
api-approvers:
  - None
creation-date: 2023-11-02
last-updated: 2023-11-02
tracking-link:
  - https://issues.redhat.com/browse/RFE-2182
  - https://issues.redhat.com/browse/OCPSTRAT-431
  - https://issues.redhat.com/browse/NE-761
see-also:
  - None
replaces:
  - None
superseded-by:
  - None
---

# Support for admin configured CA trust bundle in Ingress Operator

## Summary

This enhancement describes the proposal to allow an administrator of OpenShift

> to configure trusted Certificate Authorities(CA) to verify the service's 

Reviewer: nit
```suggestion
to configure trusted Certificate Authorities (CA) to verify the service's 
```

certificate by OpenShift Router for the re-encrypted traffic.

## Motivation

> Openshift administrator when creating a route for a service with re-encrypt termination
> type requires to provide CA certificate when a service is configured to use certificate
> signed by a CA different from the OpenShift CA to verify the service's certificate.

Reviewer: Is this what you mean to say?  The sentence is overly long...
```suggestion
When an OpenShift administrator creates a route with the re-encrypt termination
type, they have to also provide a CA certificate for that route if it uses a custom CA instead
of the default OpenShift CA for certificate validation.
```

> And when the same CA is used by the administrator for signing several services'
> certificates it is unnecessary to configure the same CA certificate content as destinationCA
> in every created route object. And is also tiresome task having to rotate the

Reviewer: ```suggestion
In addition, when the same CA is used by the administrator for signing several services'
certificates it is time-consuming and error-prone to configure the same CA certificate in every created route object. Likewise, we'd like to improve the process for updating a CA 
```

certificate when expired, compromised or for any other reason in each route object. Instead,

> if an administrator could define a ConfigMap with the list of CA certificates
> and OpenShift Ingress Controller used it along with OpenShift CA to create a common 
> CA bundle for verifying service's certificates it would make managing routes a bit simpler.

Reviewer: ```suggestion
this document proposes a simpler alternative: a feature in which an administrator could define a CA bundle for verifying certificates from a ConfigMap consisting of the list of CA certificates, associated OpenShift Ingress Controller, and trusted OpenShift CA.
```

### User Stories

> - As an OpenShift cluster administrator, I want to configure admin or the enterprise

Reviewer: What do you mean by "admin or the enterprise"?  Are these two things the same or different?

Author: Both mean the same, thought to use both the terms as put in the SDFC description. I will update it to just `admin` since it's the term used in the proposal to refer user provided CA certificate bundle.

  CA certificate bundle which should be used for verifying service's certificate by 
  the OpenShift router, instead of configuring the same as destination CA certificate 
  in each of the created route object.

> - As an OpenShift engineer, I want to create a CA trust bundle for Ingress operator usage,

Reviewer: Is this an engineer or a developer?

Author: It should be developer, I will update it.

  which includes the OpenShift service CA and the admin provided CA bundle as the default
  CA for verifying the service's certificate when a destination CA certificate is not
  configured for a route.

> - As an OpenShift engineer, I want to update the OpenShift router certificate cache whenever

Reviewer: This is probably a task only for the administrator.  Also, it may be more useful to work with expiration dates rather than updated/deleted certificates.  That way a new certificate can be added some time before the old one expires, in preparation for readiness when the new one expires.  What do you think?

Author: Yeah I agree, I thought to use updated/deleted to be generic for various reasons a certificate could be replaced. Please let me know your thoughts.

  the OpenShift service CA certificate or the admin configured CA certificate bundle is 
  updated or deleted.

### Goals

- An OpenShift administrator can configure a CA certificate bundle and expect
  that to be used for verifying the service's certificate by the OpenShift
  router without defining the destination CA certificate in the route object
  created for the service.
- Openshift Router should be updated seamlessly when the OpenShift service

>   CA certificate or the administrator configured CA certificate bundles are modified.

Reviewer: I think we need a goal to have some way to make this new feature optional. In other words we have to support both the Route configured with its own, and the Route configured to use the new trusted bundle in the ConfigMap.  We also need to define which one will be used if both are present.

Author: Both the ways will be supported, if a Route is created with destinationCA then just that will be used for verifying server certificate, else the defaultDestinationCA will be used, which will contain both the user provided CA certificates and the OpenShift CA certificate.

Reviewer: We can't have the default CA within the user-configured CA bundle.  Some routes should not be configured to have access to the user-configured CA if the route's `destinationCACertificate` is empty, that is not secure.  See https://github.com/openshift/enhancements/pull/1514/files#r1411110362

### Non-Goals

- OpenShift administrator defined CA certificate bundle will be used only for the
  traffic through OpenShift router and for the routes created with reencrypt
  termination type, any other scenario is outside the scope of this proposal.

## Proposal

OpenShift Ingress Operator to create a CA trust bundle which will contain the
OpenShift service CA certificate and the admin configured CA certificates
bundle and makes it available to the OpenShift Router as the default destination

> CA bundle file to be used for verifying the service's certificate. And also

Reviewer: How should the router associate the service to the cert, without necessarily using a field in the Route?  Not all re-encrypt routes will be using this new feature.  This is related to my comment about making sure this is optional, and defining what happens when there is both a service CA and a global CA.

Author: The user need not configure or provide a reference to the configmap in the Route object, the new CA bundle will be used by default by the router only when a destinationCA is not configured by the user.

Reviewer: Please explain why it would not be sufficient for the admin to change the default CA.  They could do the same thing by just updating the default CA with their own with their own trusted CA bundles.

Reviewer: Okay.  If you could please add a brief explanation that would be great.  Something like:  this design proposes concatenating the default CA and custom CA together for the purposes of a common CA for the route component only.  A question was raised as to why not just allow the user to concatenate them together into a default CA, and that was answered as: because this is for routes only and the default CA is used by more than just routes.

Author: comment incorporated.

Controller, and trusted OpenShift CA.

### User Stories
- As an OpenShift cluster administrator, I want to configure admin or the enterprise
  CA certificate bundle which should be used for verifying service's certificate by 
  the OpenShift router, instead of configuring the same as destination CA certificate 
  in each of the created route object.
- As an OpenShift developer, I want to create a CA trust bundle for Ingress operator usage,
  which includes the OpenShift service CA and the admin provided CA bundle as the default
  CA for verifying the service's certificate when a destination CA certificate is not
  configured for a route.
- As an OpenShift developer, I want to update the OpenShift router certificate cache whenever
  the OpenShift service CA certificate or the admin configured CA certificate bundle is 
  updated or deleted.

### Goals

- An OpenShift administrator can configure a CA certificate bundle and expect
  that to be used for verifying the service's certificate by the OpenShift
  router without defining the destination CA certificate in the route object
  created for the service.

> - Openshift Router should be updated seamlessly when the OpenShift service

Reviewer: Please define "seamlessly".  When the certificate data changes, the container needs to restart to pick up the new data, so it is not absolutely seamless.

Author: Since router updates the new certificate data automatically upon detecting changes(included as part of the proposal) and no action from user is required apart from updating the certificate bundle and reload option provided by the router for updating the certificate data doesn't require container to be restarted, I thought to mention seamlessly. Please let me know what you think.

Reviewer: I believe the router reloads (my mistake saying that the container restarts) when the backing files change, even if they are volume-mounted, but I could be wrong.  Have you observed otherwise?

Author: No, I have the same understanding. A watcher is added for the volume-mounted ca-bundle file in the Openshift Router and on detecting the changes it reloads the config.

though volume mount is updated, certain [delay](https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#mounted-configmaps-are-updated-automatically) 
is expected for the new changes to be propagated to the pod.

## Design Details

### Open Questions [optional]

N/A.

### Test Plan

- Create routes with reencrypt termination type with and with destinationCA.
- Upgrade/downgrade testing.
- Sufficient time for feedback.
- Available by default.

### Graduation Criteria

#### Dev Preview -> Tech Preview

- Ability to utilize the enhancement end to end
- End user documentation
- Sufficient test coverage
- Gather feedback from users rather than just developers

#### Tech Preview -> GA

- More testing (upgrade, downgrade, scale)
- Sufficient time for feedback
- Available by default
- Conduct load testing
- User facing documentation created in [openshift-docs](https://github.com/openshift/openshift-docs/)

#### Removing a deprecated feature
N/A.

### Upgrade / Downgrade Strategy

On upgrade:
- OpenShift Ingress Operator would have the functionality to look for `admin-ca-bundle`
  in `openshift-config` namespace and would create `ingress-ca-bundle` in

>   `openshift-ingress` namespace and OpenShift Router would refer to the new CA bundle content.

Reviewer: As it stands, this proposal now allows any route created prior to the upgrade and without a destinationCA, to now be able to access all the custom CAs that have been added.  Unless you can explain that there would be no harm to this, this is probably not acceptable.

Author: I think this shouldn't create issues, since the custom CAs are accessible to the router and will be used for verifying the service's certificate.
- If the route was accessible prior to the upgrade without a destinationCA, it would continue to be accessible and router will now have access to additional custom CAs, which will not have any impact.
- If the route was not accessible prior to the upgrade because of the reason, `router unable to verify server certificate`, it would become accessible after upgrade if the admin includes the CA used for signing the service's certificate in the bundle, else it will continue to be inaccessible.

Author: comment incorporated.

On downgrade:
- OpenShift Ingress Operator created `ingress-ca-bundle` in `openshift-ingress` namespace
  might still exist, but it will not be used by the operator or the router. And the

>   routes created post upgrade without a destinationCA will become inaccessible.

Reviewer: Do the routes become inaccessible?  They should still use the default CA after downgrade, as before.

Author: Routes which relied on the OpenShift service CA(default CA prior to upgrade) will continue to be accessible, but those routes which relied on the custom admin supplied CA certificate will become inaccessible, until user configures the destinationCACertificate for such routes.

Reviewer: Okay, please clarify this in the EP.

Author: comment incorporated.

---
title: admin-configured-ca-bundle
authors:
  - "@bhb"
reviewers:
  - "@candita" ## reviewer for cluster-ingress-operator component
  - "@Miciah" ## reviewer for cluster-ingress-operator component
  - "@tgeer" ## reviewer for CFE
approvers:
  - "@candita" ## approval from cluster-ingress-operator component
  - "@Miciah" ## approval from cluster-ingress-operator component
  - "@jerpeter1" ## approval from CFE
api-approvers:
  - None
creation-date: 2023-11-02
last-updated: 2023-12-06
tracking-link:
  - https://issues.redhat.com/browse/RFE-2182
  - https://issues.redhat.com/browse/OCPSTRAT-431
  - https://issues.redhat.com/browse/NE-761
see-also:
  - None
replaces:
  - None
superseded-by:
  - None
---

# Support for admin configured CA trust bundle in Ingress Operator

## Summary

This enhancement allows an OpenShift cluster administrator to configure
OpenShift router with custom trusted Certificate Authorities (CAs) to use to verify services' 
certificates for re-encrypted traffic. Enhancement proposes concatenating the custom and

> OpenShift service CA certificates together for the purposes of a common CA for the route component.

Author: comment incorporated.

- As an OpenShift developer, I want to create a CA trust bundle for Ingress operator usage,
  which includes the OpenShift service CA and the admin provided CA bundle as the default CA
  for verifying the service's certificate when a destination CA certificate is not configured
  for a route.
- As an OpenShift cluster administrator, I want OpenShift router to update its configuration whenever
  the OpenShift service CA certificate or the admin-configured CA certificate bundle is updated
  or deleted.

### Goals

- An OpenShift administrator can configure a CA certificate bundle and expect OpenShift router to use it
  for verifying a service's certificate when
  `destinationCACertificate` isn't specified in the route object created for the service.
- Openshift router should seamlessly reload the OpenShift service CA certificate or
  the administrator-configured CA certificate bundle when either is modified.

### Non-Goals

- OpenShift administrator defined CA certificate bundle will be used only for the traffic
  through OpenShift router and for the routes created with re-encrypt termination type, any
  other scenario is outside the scope of this proposal.

> - Openshift Ingress operator created CA bundle by concatenating the custom and OpenShift service 

Reviewer 3: I see. I find this a bit confusing. The way this is written, isn't this just a goal?

Some folks write non-goals in the negative form, or explicitly self described as out-of-scope (just like your first goal), and they should be taken at face value, like what you just described in your comment. Here's an example:
https://github.com/openshift/enhancements/blob/57d4c91edc7628fce3d54d277625f37857946700/enhancements/ingress/client-tls.md#L49-L54

Or, some folks write them in the positive, and it should be interpreted negatively as a non-goal (it is _not_ a goal):
https://github.com/openshift/enhancements/blob/b070b03e7b60533afaf93b1fe4f085796acd3717/enhancements/ingress/aws-elb-idle-timeout.md#L55-L58

As this is written, it's hard for me to interpret if it's intended to be taken at face value. I think your comment does a good job of making it clear:
```suggestion
- The common CA bundle created by the operator is for all ingress controllers usage, and defining a CA bundle specific for each IngressController usage is out of scope.
```

> - Openshift Ingress operator created CA bundle by concatenating the custom and OpenShift service 

Author: comment incorporated.

> - Openshift Ingress operator created CA bundle by concatenating the custom and OpenShift service 

Reviewer 3: I don't see any change, are you planing to push?

> - Openshift Ingress operator created CA bundle by concatenating the custom and OpenShift service 

Reviewer 3: All good, thanks for the update

  - "@candita" ## approval from cluster-ingress-operator component
  - "@Miciah" ## approval from cluster-ingress-operator component
  - "@jerpeter1" ## approval from CFE
api-approvers:
  - None
creation-date: 2023-11-02
last-updated: 2023-12-06
tracking-link:
  - https://issues.redhat.com/browse/RFE-2182
  - https://issues.redhat.com/browse/OCPSTRAT-431
  - https://issues.redhat.com/browse/NE-761
see-also:
  - None
replaces:
  - None
superseded-by:
  - None
---

# Support for admin configured CA trust bundle in Ingress Operator

## Summary

This enhancement allows an OpenShift cluster administrator to configure OpenShift router with custom
trusted Certificate Authorities (CAs) to use for verifying services' certificates of re-encrypt routes.
The enhancement proposes configuring custom trusted Certificate Authorities (CAs) which also includes the
OpenShift service CA certificate for an IngressController to be used for certificate validation on
re-encrypt routes.

## Motivation

When an OpenShift administrator creates a route with the re-encrypt termination type, they have to also
provide a CA certificate for that route if it uses a custom CA instead of OpenShift's service CA for

> certificate validation. In addition, when the same CA is used by the administrator for signing several

Reviewer 4: and this also isn't administrator, it's "namespace editor", right?

services' certificates it is time-consuming and error-prone to configure the same CA certificate in every

> created route object. Instead, this enhancement proposes a simpler alternative: in which an administrator

Reviewer 4: this one really is administrator and specifically "cluster-administrator".

Notice that by  using the incorrect term above, this reads as though it's the same person. But when using the correct terms, it's clear that we're talking about different people.

The proposal expects the OpenShift cluster administrator to create a ConfigMap with all the required 
CA certificates, including the OpenShift service CA in the same namespace of the `IngressController` object
in which the configmap reference is provided.

Enhancement requires below modifications to the `IngressController` API.
- Add `defaultDestinationCATrust` field to `.spec` of the `ingresscontrollers.operator.openshift.io`
```golang
type IngressControllerSpec struct {
	// defaultDestinationCATrust is a reference to a configmap containing the trusted
	// Certificate Authorities (CAs) used by the Router for verifying the services'
	// certificates of re-encrypt routes. When Routes don't specify their own
	// destinationCACertificate, defaultDestinationCATrust is used.
	//
	// The configmap must contain the following key and data:
	//
	//   ca-bundle.crt: trusted CA certificates, including the OpenShift service CA.
	//
	// The user is responsible for including OpenShift service CA along with their own
	// trusted CA certificates in the referenced configmap. OpenShift service CA can be
	// found in openshift-service-ca.crt configmap in openshift-config namespace.
	//

> 	// +optional

Reviewer 4: describe what happens when unset

> 	DefaultDestinationCATrust *corev1.LocalObjectReference `json:"defaultDestinationCATrust,omitempty"`

Reviewer 4: Define your own reference type (see the doc on corev1.ObjectReference), but otherwise the API field makes sense.  Since we know ClusterTrustBundles are coming, nest under a discriminated union

```go
DefaultDestinationTrust DefaultDestinationTrust

type DefaultDestinationTrust struct{
  // currently only "", ConfigMap, but later probably ClusterTrustBundle or SignerName
  Type string

  ConfigMap *corev1.LocalObjectReference `json:"defaultDestinationCATrust,omitempty"`
}
```

title: admin-configured-ca-bundle
authors:
  - "@bhb"
reviewers:
  - "@candita" ## reviewer for cluster-ingress-operator component
  - "@Miciah" ## reviewer for cluster-ingress-operator component
  - "@tgeer" ## reviewer for CFE
approvers:
  - "@candita" ## approval from cluster-ingress-operator component
  - "@Miciah" ## approval from cluster-ingress-operator component
  - "@jerpeter1" ## approval from CFE
api-approvers:
  - None
creation-date: 2023-11-02
last-updated: 2023-12-06
tracking-link:
  - https://issues.redhat.com/browse/RFE-2182
  - https://issues.redhat.com/browse/OCPSTRAT-431
  - https://issues.redhat.com/browse/NE-761
see-also:
  - None
replaces:
  - None
superseded-by:
  - None
---

# Support for admin configured CA trust bundle in Ingress Operator

## Summary

This enhancement allows an OpenShift cluster administrator to configure OpenShift router with custom
trusted Certificate Authorities (CAs) to use for verifying services' certificates of re-encrypt routes.
The enhancement proposes concatenating the custom and OpenShift service CA certificates together for the
purposes of a common CA to be used for certificate validation on re-encrypt routes.

## Motivation

> When an OpenShift administrator creates a route with the re-encrypt termination type, they have to also

Reviewer 4: doesn't a namespace editor create the route?

Author: Sorry, I am not sure what `namespace editor` is. But this is for the routes created by an admin to expose a deployed service.

Reviewer 4: > Sorry, I am not sure what `namespace editor` is. But this is for the routes created by an admin to expose a deployed service.

routes are created by low-privileged users of a cluster, not the cluster-admin.  Those low privileged users include editors who are allowed to create pods, services, and routes, but lack visibility into cluster configuration.

Author: Sorry, I used `OpenShift administrator` in generic way as those who have access to cluster and have sufficient permissions to provision required resources and I didn't think it in line with the roles.

Reviewer 4: Still outstanding.  We have different roles for our product and this is a low-privilege operation.  The openshift administrator is a high privilege account and who performs an action is considered in places in our API.  Please use the thing you mean.  I think you mean "namespace editor"

provide a CA certificate for that route if it uses a custom CA instead of OpenShift's service CA for
certificate validation. In addition, when the same CA is used by the administrator for signing several
services' certificates it is time-consuming and error-prone to configure the same CA certificate in every

> created route object. Instead, this enhancement proposes a simpler alternative: in which an administrator

Reviewer 4: ClusterTrustBundle would be  used if it was available.

- As an OpenShift cluster administrator, I want OpenShift router to verify services' certificates with just
  the configured destination CA Certificate and not the default destination CA certificate.

### Goals

- An OpenShift administrator can configure a CA certificate bundle and expect OpenShift router to use it
  for verifying a service's certificate when destination CA Certificate isn't specified in the route object
  created for the service.
- Openshift router should seamlessly reload the OpenShift service CA certificate or the
  administrator-configured CA certificate bundle when either is modified.

### Non-Goals

- OpenShift administrator defined CA certificate bundle will be used only for the traffic through OpenShift
  router and for the routes created with re-encrypt termination type, any other scenario is outside the
  scope of this proposal.
- The common CA bundle created by the operator is for all ingress controllers usage, and defining a CA
  bundle specific for each IngressController usage is out of scope.

## Proposal

Change OpenShift Ingress Operator to:
1. Create a CA trust bundle which will contain the OpenShift service CA certificate and the custom

>    user-supplied CA certificates.

Reviewer 4: further up the discussion is of cluster-admin.  Is this a different user or a cluster-admin.  if cluster-admin use the same noun please.

Reviewer 4: > It could be any user who has access to create the `admin-ca-bundle` ConfigMap in the `openshift-config` namespace, not in particular one with cluster-admin role assigned. Please let me know if it should be changed to `cluster-admin supplied`.

Which user other than a cluster-admin could do this in a stock cluster?

Author: Users with `admin` or `edit` roles could also create ConfigMaps in a given project.

> 2. Make the new common certificate available to the OpenShift Router as the default destination CA bundle

Reviewer 4: if it is the default, how would system trust be used?

Author: The default destination CA currently contains just the CA certificate used for signing openshift's own services, which will be extended to include user-supplied ca-trust as well.

   file to be used for verifying the services' certificate.
3. Updates the CA bundle file whenever either of the certificates change.

Update router to:
1. Watch the common CA bundle and update its configuration internally when needed.

### Workflow Description

- An OpenShift administrator configures the list of trusted CA certificates as a ConfigMap, named

>   specifically `admin-ca-bundle` in `openshift-config` namespace.

Reviewer 4: Instead of using magic names, use a configuration field to reference a configmap name in the openshift-config namespace.  This pattern is congruent to fields like apiserver.config.openshift.io spec.clientCA

Author: My thinking is to have something like system trust bundle for openshift defined by user, which can be used for verifying the services' certificates signed by one of the CA in bundle, instead of having to configure it in every required place. 

Route API has an option to configure the destinationCA to be used for verifying the service certificate, but instead if we can have a user ca-trust which could be used for verification, no need to configure it in each route object.

- OpenShift Ingress Operator creates `ingress-ca-bundle` ConfigMap in `openshift-ingress` namespace by

>   concatenating the admin configured CA certificates and the OpenShift service CA certificate stored in

Reviewer 4: This looks improper.  If a cluster-admin is saying, "verify using this CA bundle", then it should use "this CA bundle".  not "this CA bundle and the other ones I have".  In other kube binaries and flags, specifying a CA bundle means, "use exactly this one".

Author: Here we are combining the openshift's own CA certificate and the user provided ca bundle as common ca trust to verify the service's certificate behind a route, which is used like system ca-trust bundle. Please correct me.

  the respective ConfigMaps.
- OpenShift Ingress Operator creates a watcher and reconciles whenever changes are observed in the
  aforementioned ConfigMap objects.
- The Ingress Operator configures OpenShift router with a ConfigMap volume mount to use the operator-created ConfigMap as the default destination CA certificate
  to verify a service's certificate when a route with re-encrypt termination type is created with an empty
  destination CA.
- OpenShift router watches the volume mount and reloads HAProxy whenever the CA certificate bundle changes.

![alt Workflow diagram](admin-configured-ca-bundle.png)

> #### Variation [optional]

Reviewer 4: having reached this point.  if https://github.com/kubernetes/api/blob/39c80441f8144164dc6c2b815cbeb8564ed31e6d/certificates/v1alpha1/types.go#L28-L43 were enabled, would we be talking about having a route indicate, "use this clustertrustbundle.spec.signerName to verify me" on a field and the cluster-admin (or other) would managed the ClusterTrustBundle?

Author: Please correct me, if my understanding of the API is correct, this would require the user to make use of CertificateSigningRequests for obtaining the services' certificates, but for the routes usecase, the signer need not be made available in openshift and the service can have key pair assigned which were generated externally correct. 
If `ClusterTrustBundle` would be used, then it would be signer-unlinked types I think.

OpenShift Router should create a file watcher to observe `/var/run/configmaps/ca-trust/ca-bundle.crt`
file and reload when the file content has been modified.

### Risks and Mitigations

CA certificates configured could contain invalid PEM data or has duplicate/expired/revoked certificates
or uses insecure certificate algorithm which could pose risk and cause vulnerability.
  - CA certificates could be validated and those not meeting the requirements and could potentially disrupt
    other services' accessibility can be dropped. And the verification of services' certificates signed
    by the unconsidered certificates would fail and such services will be inaccessible.

### Drawbacks

When a user updates the `ingress-ca-bundle` ConfigMap, its contents are made available to OpenShift Router
are eventually updated, though a certain [delay](https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#mounted-configmaps-are-updated-automatically)
is expected for the new changes to be propagated to the pod.

## Design Details

### Open Questions [optional]

N/A.

### Test Plan

- Create routes with re-encrypt termination type with and without destinationCA.

>   - Routes with empty destinationCA should be accessible only

Reviewer 3: *nit* This wording and structure feels a bit vague to me, but I assume you are saying `routes with these types of services`, is that what you mean?
```suggestion
  - Routes with empty destinationCA should be accessible only when using:
      - A Service that uses a certificate signed by OpenShift CA
      - A Service that uses a certificate signed by custom CA and is present in `admin-ca-bundle`.
```

Author: I had to rebase commit to remove sentences from older commit.

	// certificates of re-encrypt routes. When Routes don't specify their own
	// destinationCACertificate, defaultDestinationCATrust is used.
	//
	// The user is responsible for including OpenShift service CA along with their own
	// trusted CA certificates in the referenced object. OpenShift service CA can be
	// found in openshift-service-ca.crt configmap in openshift-config namespace.
	//
	// If unset, OpenShift service CA configured in openshift-service-ca.crt configmap will
	// be used as default destination CA certificate.
	//
	// +optional
	DefaultDestinationCATrust DefaultDestinationTrust `json:"defaultDestinationCATrust,omitempty"`
}

// DefaultDestinationTrust specifies the reference to an object containing the trusted CA certificates.
// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'ConfigMap' ?  has(self.configMap) : !has(self.configMap)",message="configMap must be configured when type is ConfigMap"
// +union
type DefaultDestinationTrust struct {
	// type is the referenced object type which contains the trusted CA certificates.
	// The default value is ConfigMap. ConfigMap is the only supported type at present.
	//
	// +unionDiscriminator

> 	// +kubebuilder:default:="ConfigMap"

Reviewer 4: I suspect you'll need "" without defaulting to handle the, "I'm using the default case for NE today"

## Proposal

OpenShift Ingress Operator to create a CA trust bundle which will contain the

> OpenShift service CA certificate and the admin ed CA certificates

Reviewer: Is this what you mean?
```suggestion
default OpenShift service CA certificate and the custom user-supplied CA certificates
```

Author: suggestion incorporated.

bundle and makes it available to the OpenShift Router as the default destination
CA bundle file to be used for verifying the service's certificate. And also
updates the CA bundle file whenever either of the certificates changes, which
should be watched by the OpenShift Router and update the same internally.

### Workflow Description

> - An OpenShift administrator es the list of trusted CA certificates

Author: Yeah, I think, not sure how it got changed during update.
I have corrected in other places too.
